src/Supervisor.cpp

```diff
@@ -1221,4 +1221,63 @@ ZunResult Supervisor::FadeOutMusic(f32 fadeOutSeconds)
     return ZUN_SUCCESS;
 }
 #pragma optimize("", on)
+
+#pragma optimize("s", on)
+// this is for rebinding keys
+u8 *th06::Controller::GetControllerState()
+{
+    static u32 controllerData[32];
+    memset(&controllerData, 0, sizeof(controllerData));
+    if (g_Supervisor.controller == NULL)
+    {
+        // TODO: not tested
+        JOYINFOEX joyinfo;
+        memset(&joyinfo, 0, sizeof(JOYINFOEX));
+        joyinfo.dwSize = sizeof(JOYINFOEX);
+        joyinfo.dwFlags = JOY_RETURNALL;
+        MMRESULT MVar1 = joyGetPosEx(0, &joyinfo);
+        if (MVar1 == 0)
+        {
+            u32 local_3c = joyinfo.dwButtons;
+            for (u32 local_40 = 0; local_40 < 32; local_40 += 1)
+            {
+                if ((local_3c & 1) != 0)
+                {
+                    *(u8 *)((int)controllerData + local_40) = 0x80;
+                }
+                local_3c = local_3c >> 1;
```

Comment:
Another thing ZUN likes doing and ghidra just doesnt decompiles well, its multiple loop expressions inside a `for`
The for its like this:
```c
for (u32 local_3c = joyinfo.dwButtons, u32 local_40 = 0; local_40 < 32; local_40 += 1, local_3c >>= 1)
```
but ghidra just puts the second loop expression at the end of the for